### ISSUE TITLE
Add removing of tags

### ISSUE THREAD

**Issue by thekk1** [nextcloud/files_automatedtagging#649]
In some cases it is needed to remove tags from files.
Please add this feature to this application.
Thx.

**Comment by nickvergessen** [comment]
Can you describe how this should work? The app currently mostly acts on uploading a new file. So in what case would you like to remove a tag?

**Comment by thekk1** [comment]
There are many rules you can use so in this generic way it is a regular use case to remove what you set, isn't it?

In my specific case, an other app set tags and don't provide a way to remove them. So I want to set a rule that removes a tag from a file that is hit by the conditions.
Sure, this is a workaround and the origin app has maybe to provide a way.

Other use cases can be based on my initially generic description: You set a tag based on conditions and want to have a rule to remove the tags in other direction.
It's the third time that I stumbled over the fact that tags can only be removed manually by hand. Bad thing if thousands of files where tagged.

How this should work?
Why don't make a select field for the action that should be done?
Options could be:
- Set Tag
- Remove Tag

**Comment by nickvergessen** [comment]
> So I want to set a rule that removes a tag from a file that is hit by the conditions.

The problem is the rules are only executed on uploading a file. Are you actually uploading a file?

I guess you would need something else like a "live execution" which then crawls over all your files and removes the tag not matching the conditions? However apart from the tagging (which is literally a single line of this apps code https://github.com/nextcloud/files_automatedtagging/blob/master/lib/Operation.php#L94 ) it has very little in common with this app. The conditions come from the workflowengine app https://github.com/nextcloud/server/tree/master/apps/workflowengine

So I'm not sure it makes sense to combine this different requirements into a single app.

**Comment by thekk1** [comment]
> > So I want to set a rule that removes a tag from a file that is hit by the conditions.
> 
> The problem is the rules are only executed on uploading a file. Are you actually uploading a file?
As I said, this is not true. The condition is "file changed". 

> I guess you would need something else like a "live execution" which then crawls over all your files and removes the tag not matching the conditions?
No. As I said, I need a way to remove a tag if a condition is true. As far as I'm concerned, it could also be something else like an shell script. I could setup a script with an SQL statement for the database to remove the tag but I realy don't want to go this way.

> However apart from the tagging (which is literally a single line of this apps code https://github.com/nextcloud/files_automatedtagging/blob/master/lib/Operation.php#L94 ) it has very little in common with this app. The conditions come from the workflowengine app https://github.com/nextcloud/server/tree/master/apps/workflowengine
> 
> So I'm not sure it makes sense to combine this different requirements into a single app.

I admit that I don't know how the NC framework is sticked together but I can not believe that is so hard to make a field in the setup view to serve the condition in workflow part and use it also as a field for the action to do. Maybe this is not needed if you just check if there is a tag condition choosed and based on that you provide the option to remove instead of set the tag.

**Comment by nickvergessen** [comment]
> The condition is "file changed".

That's right, but that is only on uploading, not when it was tagged. As long as that matches your flow it should be doable yes.

Mind to bring in some light which other app is applying the tag that you want to remove? This would allow us to test locally that it really does work. Because otherwise it could be that the other app is running after autotagging and reapplying the tag that we just removed with the rule.

**Comment by thekk1** [comment]
> > The condition is "file changed".
> 
> That's right, but that is only on uploading, not when it was tagged. As long as that matches your flow it should be doable yes.

Now I am really confused. Why do you relate the condition we talking about  to "when it was tagged"?
For clearness this is what I understood:
The rule that I configure, no matter what conditions I set or if they are matches or not, will be applied when a file is changed. This can be an upload, create, copy, move or simple change content of the file.
Is there anything wrong or do I missed something? Is tagging a file included in "file changed"? I guessed a tag is just a relating property on the database.

What I asked for is really simple:
I want to remove a tag if a file is changed (any case up described, for sure it does not make sense on create or upload) and the conditions are matched (a filename pattern, a specific tag existing at the file).

> Mind to bring in some light which other app is applying the tag that you want to remove? This would allow us to test locally that it really does work. Because otherwise it could be that the other app is running after autotagging and reapplying the tag that we just removed with the rule.

In this case, I talk about the app "approval". There are no automated tagging rules as far as I understood.
The app just uses tags (invisible tags) to mark a state in the approval process which can be "process started", "approved" and "not approved". It's not possible to restart the process after the state is "approved" or "not approved". The only way is to remove the tag manually which makes the admin quitting his job or compromises the process  if you change the tags from invisible to normal so users can do it by themself.

Another case, far ago, was:
I set tags on files that was edited by guests and want to remove the tag after the file was moved or changed by a user from a specific group.

**Comment by nickvergessen** [comment]
That's all okay and now I understand the problem.
I would still recommend to report this at the approval app. Might be quicker to solve there and that would make everyone benefit from the adjustment.

Nevertheless we can have this issue here additionally to allow working around the problem manually.

**Comment by thekk1** [comment]
The case is already reported to the app developers.
I don't asked for a solution to this specific problem, I asked in general for a way to remove tags by conditions because I stumbled multiple times over the fact that tags are not removable in an automatic way.

**Comment by madscientist42** [comment]
It rather is a succinct problem when you pause to think about it.  

It's not just something that is an inconsistency with the approval app, it should be a swiss army knife here in the context of things.  If you're saying "file changed" that means uploaded/reuploaded, etc. at which point in time, I'd like to see it see the set approvals tags and wipe them so that one starts fresh with each version like any sensible versioning control system.

Otherwise, this is...well...of somewhat limited use, really.  You can't realistically make workflows without an ability to rules remove tags if present on a given file/directory as well as set them.

**Comment by madscientist42** [comment]
I will state that it's a bit concerning that it's been nearly a year since the last exchanges on this issue...  Enhancement IS a good description and it DOES make sense...any feedback from the team there on what the plan/status is on this one?  It does empower making real flows work properly for re-cocking initial conditions, etc. so the other apps do the right things.

**Comment by Mululu** [comment]
Is there any news about this? Without this function, the app is useless because if a text document has already been released and is then changed, the status must be released again and the released status must be removed again.

**Comment by lekhnath** [comment]
I want a tag (Approved) added by my approval workflow to be removed when the file content is changed. This is a must have option IMO.

**Comment by Ceelight** [comment]
I would also love to see that tags can be removed. Let me give you a use case:

We would like to classify our documents according to the "Traffic Light Protocol". The idea is to check whether the string "TLP-AMBER" appears in the file name, for example. If this is the case, the tag "TLP:AMBER" is assigned. If any "TLP-*" is missing, the tag "No TLP!" is assigned. The employee now has the option of renaming the file to comply with internal regulations. This would technically be a change to the file so that the flow checks start again. "No TLP!" should now be removed again.

And this is exactly what is currently not possible.

**Comment by MalbrottsAsnan** [comment]
This seems like a highly sought after feature, and it would greatly help with what I'm trying to accomplish, which is to respect both a file's modification date and creation date when using Automated Tagging in combination with File Retention. The most intuitive way to add this would in my opinion be to add a "remove tag if present" option in the flow rule configuration. The logic for triggering the rule and its conditions can remain the same, only the executed action needs to be modified. Is this possible and would it be complicated to develop?

**Comment by rumbertl** [comment]
I solved this with the https://apps.nextcloud.com/apps/workflow_script app.

The script uses this command:

(in docker)
php /var/www/html/occ tag:files:delete "$1" "$2 "$3"

in the Flow-Config I wrote:

/path/to/script.sh %i TAGNAME public

**Comment by madscientist42** [comment]
> I solved this with the https://apps.nextcloud.com/apps/workflow_script app.
> 
> The script uses this command:
> 
> (in docker) php /var/www/html/occ tag:files:delete "$1" "$2 "$3"
> 
> in the Flow-Config I wrote:
> 
> /path/to/script.sh %i TAGNAME public

Kinda clumsy.  Be better to have either the plugin have a yank mode or to get files_scripts back fully running- as it will do that MUCH cleaner.
